homeassistant/components/icloud/config_flow.py

```diff
@@ -0,0 +1,224 @@
+"""Config flow to configure the iCloud integration."""
+import logging
+import os
+
+import voluptuous as vol
+from pyicloud import PyiCloudService
+from pyicloud.exceptions import PyiCloudException, PyiCloudFailedLoginException
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
+from homeassistant.util import slugify
+
+# pylint: disable=unused-import
+from .const import (
+    CONF_ACCOUNT_NAME,
+    CONF_GPS_ACCURACY_THRESHOLD,
+    CONF_MAX_INTERVAL,
+    DEFAULT_MAX_INTERVAL,
+    DEFAULT_GPS_ACCURACY_THRESHOLD,
+    DOMAIN,  # noqa
+)
+
+CONF_TRUSTED_DEVICE = "trusted_device"
+CONF_VERIFICATION_CODE = "verification_code"
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class IcloudFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a iCloud config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
+
+    def __init__(self):
+        """Initialize iCloud config flow."""
+        self.api = None
+        self._username = None
+        self._password = None
+        self._account_name = None
+        self._max_interval = None
+        self._gps_accuracy_threshold = None
+
+        self._trusted_device = None
+        self._verification_code = None
+
+    def _configuration_exists(self, username: str, account_name: str) -> bool:
+        """Return True if username or account_name exists in configuration."""
+        for entry in self._async_current_entries():
+            if (
+                entry.data[CONF_USERNAME] == username
+                or entry.data.get(CONF_ACCOUNT_NAME) == account_name
+                or slugify(entry.data[CONF_USERNAME].partition("@")[0]) == account_name
+            ):
+                return True
+        return False
+
+    async def _show_setup_form(self, user_input=None, errors=None):
+        """Show the setup form to the user."""
+
+        if user_input is None:
+            user_input = {}
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(
+                        CONF_USERNAME, default=user_input.get(CONF_USERNAME, "")
+                    ): str,
+                    vol.Required(
+                        CONF_PASSWORD, default=user_input.get(CONF_PASSWORD, "")
+                    ): str,
+                }
+            ),
+            errors=errors or {},
+        )
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initiated by the user."""
+        errors = {}
+
+        icloud_dir = self.hass.config.path("icloud")
```

Comment:
We should use `storage.Store`

Reply:
Or actually, we should just store things in the config entry data.

Reply:
I don't think we can make it using the config entry data  following `pyicloud`'s code (sadly) : https://github.com/picklepete/pyicloud/blob/9bb6d750662ce24c8febc94807ddbdcdf3cadaa2/pyicloud/base.py#L136-L158

Will investigate on `storage.Storage`

Reply:
Tried and having this, don't know how to do.

config_flow.py
<img width="995" alt="Capture d’écran 2019-12-02 à 13 43 11" src="https://user-images.githubusercontent.com/14821482/69960455-cd365380-1509-11ea-8e2e-6796a9226c6a.png">
__init__.py
<img width="1025" alt="Capture d’écran 2019-12-02 à 13 42 41" src="https://user-images.githubusercontent.com/14821482/69960462-d0c9da80-1509-11ea-9492-1b6d166625f6.png">

Tests are passing

Reply:
The stack trace on your screenshot is because you're trying to initialize the Store inside `setup`. You can only initialize it inside an async function.

Reply:
Indeed, working nice 👍 

iCloud cookie stored in `.storage/icloud` now.